Implemented C# GC get heap size/ get used size

jeanclaudegrenier opened the pull request:
Would love some input regarding the tests.
They are very approximative, but should be stable. I can't unfortunately check for a precise size, as far as I know.

Edit: turns out they are not as stable as I thought... thanks Yamato 😓 .
Does anyone have a suggestion for a better test?

jeanclaudegrenier committed:
Implemented C# GC get heap size/ get used size

joncham commented on unity/UnityEmbedHost.Tests/BaseEmbeddingApiTests.cs:
```diff
@@ -783,6 +783,30 @@ public void UnityMethodIsGenericReturnsProperValue(Type objType, string methodNa
+        int[] data = new int[1024 * 1024 * 100];
+        GC.Collect();
+        long heapSizeAfterBigAlloc = ClrHost.gc_get_heap_size();
+        Assert.Greater(heapSizeAfterBigAlloc, heapSize);
```
Try putting `GC.KeepAlive(data);` at the end of the method.

jeanclaudegrenier committed:
Added a GC.KeepAlive as suggested in PR to stabilize the test.

jeanclaudegrenier committed:
Forcing a synchronous aggressive GC first, as we are currently asserting that GC was not ready

jeanclaudegrenier committed:
Modified test to validate that the heap/used size is more than the dummy data allocated, and that used size is smaller than the total heap

jeanclaudegrenier committed:
Added a timeout and a sleep to the gc test, to wait for the first gc to finish.

jeanclaudegrenier committed:
Trying to add a GC.WaitForPendingFinalizers call, and a try catch to figure out what exception occured

jeanclaudegrenier committed:
boosted timeout to 50 seconds as 5 seems too short in some cases. modified gc_get_used_size to use GC.GetGCMemoryInfo as there seems to be random discrepencies between GC.GetGCMemoryInfo values and GC.GetTotalMemory data (it may also be a race condition issue)

jeanclaudegrenier committed:
Removed the test's try/catch (the exception was just the timeout being triggered)
Removed test to check that used memory is less or equal to total memory, since this is likely to cause instabilities if the GC kicks in between the calls, as it seems like it happened in prior tests.

jeanclaudegrenier commented:
Ok finally got everything stable (fingers crossed) tests passed on the last 3 runs (aside from one build timeout, unrelated).
